Add the SA-IS algorithm

Add Doxygen to the project to build documentation

Add python 2, python3, perl, lua, java build targets if possible.

Add skiplists template, then convert sparse-bitpacked-array to C++ and use the sparse lists

Add highly optimized perfect alignment algorithm

Add a 2D matrix which uses skiplists in rows and columns

Re-evaluate Upper-diagonal-square-matrix functions to make index
correlation more regular, make getSize()  a const_expr if applicable.

Incorporate useful parts of PEPIS
  <http://bioinfo.noble.org/PolyGenic_QTL/Download.gy>

Add GEO .soft file tests

Add GEO .soft file helper functions to extract information related to GDS, GSM,
GPL subtypes.

Add benchmarking portions in test/

Evaluate and incorporate the following repsoitories:
<http://www.biostat.jhsph.edu/~hji/cisgenome/index_files/manual.htm>
<https://www.sanger.ac.uk/science/tools/caf>
<https://github.com/ComparativeGenomicsToolkit/cactus>
<https://github.com/fukunagatsu/CapR>
<https://code.google.com/archive/p/cbrowse/>
<https://code.google.com/archive/p/campways/>
<https://sourceforge.net/projects/chipmeta/files/?source=navbar>
<https://github.com/voges/calq>
<http://www.ccp4.ac.uk/>
<https://github.com/KarchinLab/CHASMplus>
<http://ruby.chemie.uni-freiburg.de/~martin/chemtool/>
<http://emboss.open-bio.org/rel/rel6/apps/cacheensembl.html>
<http://emboss.open-bio.org/rel/rel6/apps/cachedas.html>
<https://sourceforge.net/projects/wgs-assembler/files/?source=navbar>
<http://www.htslib.org/>
<http://clark.cs.ucr.edu/>
<https://github.com/trinityrnaseq/trinityrnaseq/wiki>
<https://github.com/mcveanlab/mccortex>
<http://abacus.gene.ucl.ac.uk/software/paml.html>
<https://sourceforge.net/projects/chipotle-2/files/chipotle-2/ChIPOTle2.0/>
<http://mrfast.sourceforge.net/>
<http://bioinf.gen.tcd.ie/~faresm/software/software.html#caps2>
<http://bioinfo.cnio.es/index.php/services/bioinfotools/22-cargo-cancer-and-related-genes-online>
<http://boevalab.com/FREEC/>
<https://sourceforge.net/projects/coperead/>
<https://github.com/kaibader/CaSSiS>
<https://www.openms.de/>
<https://github.com/tingchenlab/CROP>
<https://simtk.org/projects/campaign>
<https://sourceforge.net/projects/camino-trackvis/files/0.2/>
<https://sourceforge.net/projects/camino-trackvis/files/ERN/>
<https://sourceforge.net/projects/kakscalculator2/>
<http://www.xavierdidelot.xtreemhost.com/clonalframe.htm?i=2>
<https://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastDocs&DOC_TYPE=Download>
<https://github.com/abhatta3/Condition-dependent-Correlation-Subgroups-CCS>
<https://bitbucket.org/Wenan/caviarbf/src/7e428645be5eb518eebf8df18b436c3eca898ca2/src/?at=default>
<https://github.com/cambDI/camb>
<https://sites.google.com/site/pierrickcoupe/softwares/denoising-for-medical-imaging/multiphoton-filtering>
<http://www.bioinformatics.org/cgi-bin/viewvc.cgi/catch/>
<http://emboss.open-bio.org/html/use/apbs04.html#EMBOSSAppsTableR6>
<http://emboss.open-bio.org/html/use/apbs05.html>
<http://emboss.open-bio.org/html/dev/ch01s02.html>
<https://github.com/Illumina/canvas>
<https://www.bsse.ethz.ch/cbg/software/ct-cbn.html>
<https://github.com/grenaud/bedBamAndBeyond/>
<https://sourceforge.net/projects/seqgenomebrowse/>
<http://silicos-it.be.s3-website-eu-west-1.amazonaws.com/software/filter-it/1.0.2/filter-it.html>
<http://noch.sourceforge.net/>
<https://github.com/hyunminkang/apigenome>
<https://github.com/GreenleafLab/motifmatchr/>
<https://github.com/labsquare/fastQt>
<https://github.com/cschlaffner/PoGo>
<https://biosails.github.io/pheniqs/>
<https://github.com/smith-chem-wisc/mzLib>
<https://sourceforge.net/projects/ncl/>
<https://github.com/kdmurray91/libqes>
<https://bitbucket.org/mzytnicki/multi-mapping-counter>
<https://sourceforge.net/projects/libgdf/>
<https://sourceforge.net/projects/htqc/>
<https://sourceforge.net/projects/lwgv/files/?source=navbar>
<https://sourceforge.net/projects/skittle/>
<https://sourceforge.net/projects/quicktsaf/files/?source=navbar>
<http://www.g-language.org/kbws/>
<http://bioinfo.single-cell.cn/rna-qc-chain.html>
<https://github.com/mirand863/gcsplit>
<http://www.bcgsc.ca/platform/bioinfo/software/laneruler>
<http://www.metatrans.org/>
<https://bitbucket.org/ball/ball>
<https://github.com/davidhwyllie/findNeighbour>
<http://sparx-em.org/sparxwiki/>
<https://github.com/vincent-picaud/Joint_Baseline_PeakDeconv>
<https://code.google.com/archive/p/ftdr/>
<http://gedas.bizhat.com/gedas.htm>
<https://github.com/imgag/ngs-bits>
<https://sourceforge.net/projects/samstat/>
<https://sourceforge.net/projects/gen3d/>
<http://mycor.nancy.inra.fr/egglib/>
<https://www2.mrc-lmb.cam.ac.uk/relion/index.php?title=Main_Page>
<https://github.com/lorinanthony/MAPIT>
<https://github.com/tsirigos/ibm-cbc-genomic-tools>
<https://github.com/wingolab-org/pecaller>
<http://www.bioinf.org.uk/software/bioplib/>
<https://github.com/3DGenomes/binless>
<https://github.com/walaj/VariantBam>
<https://github.com/LohmuellerLab/PReFerSim>
<https://github.com/FePhyFoFum/phyx>
<https://code.google.com/archive/p/libzerg/>
<http://users-cs.au.dk/cstorm/software/tqdist/>
<https://github.com/ABI-Software/ICMA>
<https://github.com/slowkow/snpsea/>
<https://github.com/MWSchmid/Rcount>
<https://github.com/danmaclean/2kplus2>
<https://cran.r-project.org/web/packages/graphkernels/index.html>
<https://sourceforge.net/projects/whiteboard-class/>
<https://github.com/terraphast/terraphast>
<https://github.com/jujubix/rnadecoder>
<https://sourceforge.net/projects/starrynite/>
<https://openstructure.org/docs/1.7.1/install/>
<https://github.com/MWSchmid/HiCdat>
<https://github.com/yarden/MISO>
<https://github.com/genome/breakdancer>
<https://github.com/szpiech/garlic>
<https://github.com/TimeDependentChipSeqAnalyser/TDCA>
<https://chaste.cs.ox.ac.uk/trac/wiki/ChasteGuides/AccessCodeRepository>
<https://sourceforge.net/projects/pysces/>
<https://bitbucket.org/tobiasmarschall/clever-toolkit>
<https://sourceforge.net/projects/cdaotools/>
<https://github.com/cosmo-team/cosmo/tree/VARI-merge>
<https://github.com/arq5x/bedtools2>
<https://sourceforge.net/projects/chip-seq/files/chip-seq/1.5.3/>
<https://github.com/osallou/cassiopee-c>
<https://github.com/gersteinlab/MUSIC>
<https://liberles.cst.temple.edu/Software/CASS/index.html#Download>
<https://sourceforge.net/p/ccpn/code/HEAD/tree/branches/stable/ccpn/>
<https://github.com/nedaz/cola>
<https://hive.biochemistry.gwu.edu/dna.cgi?cmd=censuscope>
<https://sourceforge.net/projects/xat/files/CAT/0.8.6/>
<http://emboss.open-bio.org/html/use/apbs06.html>
<https://code.google.com/archive/p/mesquitegsoc09/source/default/source>
<https://github.com/pjotrp/biopp/tree/master>
<https://github.com/samtools/samtools>
